packages/grafana-data/src/events/EventBus.ts

```diff
@@ -94,40 +96,50 @@ export class EventBusSrv implements EventBus, LegacyEmitter {
 }
 
 /**
- * @alpha
- *
  * Wraps EventBus and adds a source to help with identifying if a subscriber should react to the event or not.
  */
-export class EventBusWithSource implements EventBus {
-  source: string;
-  eventBus: EventBus;
-
-  constructor(eventBus: EventBus, source: string) {
-    this.eventBus = eventBus;
-    this.source = source;
-  }
+class ScopedEventBus implements EventBus {
+  // The path is not yet exposed, but can be used to indicate nested groups and support faster filtering
+  constructor(public path: string[], private eventBus: EventBus, private localOnly: boolean) {}
 
   publish<T extends BusEvent>(event: T): void {
-    const decoratedEvent = {
-      ...event,
-      ...{ payload: { ...event.payload, ...{ source: this.source } } },
-    };
-    this.eventBus.publish(decoratedEvent);
+    if (!event.origin) {
+      (event as any).origin = this;
+    }
+    this.eventBus.publish(event);
   }
 
-  subscribe<T extends BusEvent>(eventType: BusEventType<T>, handler: BusEventHandler<T>): Unsubscribable {
-    return this.eventBus.subscribe(eventType, handler);
+  updateScope(localOnly: boolean) {
+    this.localOnly = localOnly;
   }
 
   getStream<T extends BusEvent>(eventType: BusEventType<T>): Observable<T> {
-    return this.eventBus.getStream(eventType);
+    const stream = this.eventBus.getStream(eventType);
+    if (this.localOnly) {
+      return new Observable<T>((observer) => {
+        const sub = stream.subscribe({
+          next: (evt) => {
+            if (evt.origin === this) {
+              observer.next();
+            }
+          },
+        });
+        return () => {
+          sub.unsubscribe();
+        };
+      });
+    }
+    return stream;
   }
 
   removeAllListeners(): void {
     this.eventBus.removeAllListeners();
   }
 
-  isOwnEvent(event: BusEventWithPayload<PayloadWithSource>): boolean {
-    return event.payload.source === this.source;
+  /**
+   * Creates a nested event bus structure
+   */
+  newScopedBus(key: string, localOnly?: boolean): EventBus {
+    return new ScopedEventBus([...this.path, key], this, Boolean(localOnly));
```

Comment:
not yet used, but this will get: `[dashboard, panel, legend]`

We can leave this out for now